compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -682,16 +682,304 @@ abstract class GenJSCode extends plugins.PluginComponent
             "repeated parameters.")
       }
 
-      // Implementation restriction
-      for (tree <- secondaryCtorTrees) {
-        reporter.error(tree.pos,
-            "Implementation restriction: Scala.js-defined JS classes cannot " +
-            "have secondary constructors")
+      withNewLocalNameScope {
+        val infoBuilder = new MethodInfoBuilder()
+        infoBuilder.setEncodedName("constructor")
+        infoBuilder.setIsExported(true)
+        infoBuilder.setIsStatic(false)
+
+        val infoBuilderOpt = Some(infoBuilder)
+
+        val primaryCtor =
+          genMethodWithInfoBuilder(primaryCtorTree, infoBuilderOpt).get._1
+        val secondaryCtors = secondaryCtorTrees.map { tree =>
+          genMethodWithInfoBuilder(tree, infoBuilderOpt).get._1
+        }
+
+        val dispatch =
+          genJSConstructorExport(constructorTrees.map(_.symbol), infoBuilder)
+
+        val isConstructor =
+          secondaryCtors.map(_.name.name).toSet + primaryCtor.name.name
+
+        val constructorGraph =
+          mkConstructorGraph(primaryCtor, secondaryCtors, isConstructor)
+
+        val overloadIdent = js.Ident("overload")
+
+        val js.MethodDef(_, dispatchName, dispatchArgs, dispatchResultType,
+            dispatchResolution) = dispatch
+
+        val overloadSelection = mkOverloadSelection(constructorGraph,
+            overloadIdent, dispatchResolution, isConstructor)
+
+        val preSuperCall =
+          constructorGraph.mkPreSuperCall(overloadIdent, dispatch, isConstructor)
+
+        val superCall = primaryCtor.body
+
+        val postSupperCall =
+          constructorGraph.mkPostSuperCall(overloadIdent, isConstructor)
+
+        val newBody = js.Block(overloadSelection :::
+            preSuperCall :: superCall :: postSupperCall :: Nil)
+
+        val newConstructor = js.MethodDef(static = false, dispatchName, dispatchArgs,
+            jstpe.NoType, newBody)(dispatch.optimizerHints, dispatch.hash)
+
+        // Boldly get rid of everything in infoBuilder, because it's broken
+        currentClassInfoBuilder.addMethod(
+            ir.Infos.generateMethodInfo(newConstructor))
```

Comment:
Could we get rid of the InfoBuilders first? It seems to be a rather easy change.